### PR TITLE
Canonical ordering tests

### PR DESCRIPTION
**Pull request by mkv-vcm**
Tests that I hope cover most of the error cases with canonical ordering in place now.

### CODE REVIEW

File: src/parsing/transaction.ts [hunk 1]
```diff
@@ -90,8 +90,8 @@ function parseTokenBundle<T>(tokenBundle: AssetGroup[], emptyTokenBundleAllowed:
     validate(policyIds.length == new Set(policyIds).size, InvalidDataReason.MULTIASSET_INVALID_TOKEN_BUNDLE_NOT_UNIQUE)
 
     // enforcing of policies order is removed for now and will be added back after the ordering is properly defined by a CIP
```
Comment (janmazak) [inline src/parsing/transaction.ts:92]: This comment should go away.

File: src/parsing/transaction.ts [hunk 2]
```diff
@@ -71,11 +71,11 @@ function parseAssetGroup<T>(assetGroup: AssetGroup, parseTokenAmountFn: ParseTok
     validate(assetNamesHex.length == new Set(assetNamesHex).size, InvalidDataReason.MULTIASSET_INVALID_ASSET_GROUP_NOT_UNIQUE)
 
     // enforcing of asset order is removed for now and will be added back after the ordering is properly defined by a CIP
```
Comment (janmazak) [inline src/parsing/transaction.ts:73]: to be removed.

File: src/parsing/transaction.ts [hunk 3]
```diff
@@ -90,8 +90,8 @@ function parseTokenBundle<T>(tokenBundle: AssetGroup[], emptyTokenBundleAllowed:
     validate(policyIds.length == new Set(policyIds).size, InvalidDataReason.MULTIASSET_INVALID_TOKEN_BUNDLE_NOT_UNIQUE)
```
Comment (janmazak) [inline src/parsing/transaction.ts:90]: I think we do not need this anymore?

File: src/parsing/transaction.ts [hunk 4]
```diff
@@ -71,11 +71,11 @@ function parseAssetGroup<T>(assetGroup: AssetGroup, parseTokenAmountFn: ParseTok
     validate(assetNamesHex.length == new Set(assetNamesHex).size, InvalidDataReason.MULTIASSET_INVALID_ASSET_GROUP_NOT_UNIQUE)
```
Comment (janmazak) [inline src/parsing/transaction.ts:71]: Replaced by the ordering requirement? (I don't think the error of duplicit entries will be common, so we can get rid of the MULTIASSET_INVALID_ASSET_GROUP_NOT_UNIQUE exception and reduce validation to just the ordering.

File: src/parsing/transaction.ts [hunk 5]
```diff
@@ -119,6 +119,7 @@ export function parseTransaction(tx: Transaction): ParsedTransaction {
     const certificates = parseCertificates(tx.certificates ?? [])
 
     // withdrawals
+    // we can't check here, but withdrawal keys (the StakeCredential) should be in canonical ordering
```
Comment (janmazak) [inline src/parsing/transaction.ts:122]:
```suggestion
    // we can't check here, but withdrawal map keys (derived from stake credentials) should be in CBOR canonical ordering
```

Comment (janmazak) [reply src/parsing/transaction.ts:122]: The comments on "Transaction" should be updated (i.e. in all three places where we require canonical ordering). See
```
    /**
     * Mint or burn instructions (if any).
     * Assets to be minted (token amount positive) or burned (token amount negative) with the transaction.
     * If not null, the entries' keys (policyIds) must be sorted lowest value to highest to reflect a canonical CBOR.
     * The sorting rules (as described in the [CBOR RFC](https://datatracker.ietf.org/doc/html/rfc7049#section-3.9)) are:
     *  * if two keys have different lengths, the shorter one sorts earlier;
     *  * if two keys have the same length, the one with the lower value in lexical order sorts earlier.
     */
    // TODO update the docstring when CIP-0021 is approved
    mint?: Array<AssetGroup> | null,
```

File: src/parsing/transaction.ts (hunk 4 above)
Comment (mkv-vcm) [reply src/parsing/transaction.ts:71]: What would the advantage of that be? Reducing validation. The sorting doesn't solve this one.

File: src/parsing/transaction.ts (hunk 5 above)
Comment (janmazak) [reply src/parsing/transaction.ts:122]: I think it's OK not to reference a particular CIP for that (it does not have a definite number yet), just CBOR canonical RFC is OK, we probably don't need to copy its contents here.

File: src/parsing/transaction.ts (hunk 1 above)
Comment (mkv-vcm) [reply src/parsing/transaction.ts:92]: Yeah.

File: src/parsing/transaction.ts (hunk 2 above)
Comment (mkv-vcm) [reply src/parsing/transaction.ts:73]: Exactly.

File: src/parsing/transaction.ts (hunk 4 above)
Comment (janmazak) [reply src/parsing/transaction.ts:71]: Ah, I see, this version of validation via sorting (unlike the one in ledger) does not exclude duplicity. OK, let's keep it, then.

**Review by janmazak — APPROVED** [review]

**Comment by janmazak** [comment]
Merged into #130.

**Pull request closed without merging**